### PR TITLE
Introduce support for Spring's @Scheduled

### PR DESCRIPTION
**Pull request by aureamunoz**

### CODE REVIEW

File: bom/runtime/pom.xml [hunk 1]
```diff
@@ -734,6 +734,11 @@
                 <artifactId>quarkus-spring-security</artifactId>
                 <version>${project.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-spring-scheduling</artifactId>
+                <version>${project.version}</version>
+            </dependency>
```
Comment (gsmet) [inline bom/runtime/pom.xml:741]: The deployment artifact should also be added to the deployment bom.

File: extensions/spring-scheduling/deployment/pom.xml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-core-deployment</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc-deployment</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-scheduler</artifactId>
```
Comment (gsmet) [inline extensions/spring-scheduling/deployment/pom.xml:32]: You don't need that one, it will come with the deployment dependency.

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import io.quarkus.arc.deployment.AnnotationsTransformerBuildItem;
+import io.quarkus.arc.processor.AnnotationsTransformer;
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+
+/*
+ * A simple processor that maps annotations Spring DI annotation to CDI annotation
```
Comment (gsmet) [inline extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:27]:
```suggestion
 * A simple processor that maps Spring DI annotations to Quarkus annotations
```

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+            @Override
+            public void transform(TransformationContext context) {
+                MethodInfo methodInfo = context.getTarget().asMethod();
+                List<AnnotationInstance> collect = methodInfo.annotations().stream()
+                        .filter(annotationInstance -> annotationInstance.name().equals(SPRING_SCHEDULED))
+                        .collect(Collectors.toList());
+                if (!collect.isEmpty()) {
+                    context.transform()
+                            .add(QUARKUS_SCHEDULED, AnnotationValue.createStringValue("cron", "0/1 * * * * ?")).done();
```
Comment (gsmet) [inline extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:59]: I suppose you plan to not hardcode it?

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            @Override
+            public boolean appliesTo(AnnotationTarget.Kind kind) {
+                return kind == AnnotationTarget.Kind.METHOD;
+            }
+
+            @Override
+            public void transform(TransformationContext context) {
+                MethodInfo methodInfo = context.getTarget().asMethod();
+                List<AnnotationInstance> collect = methodInfo.annotations().stream()
+                        .filter(annotationInstance -> annotationInstance.name().equals(SPRING_SCHEDULED))
```
Comment (gsmet) [inline extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:55]: Is this annotation repeatable? If so, there's a good chance the repeatable container will need some special handling (and a test).

File: integration-tests/spring-scheduling/pom.xml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+<!--        <dependency>-->
+<!--            <groupId>io.quarkus</groupId>-->
+<!--            <artifactId>quarkus-spring-di</artifactId>-->
+<!--        </dependency>-->
+
+<!--        &lt;!&ndash; only needed because of @RequestScope is utilized in the tests &ndash;&gt;-->
+<!--        <dependency>-->
+<!--            <groupId>io.quarkus</groupId>-->
+<!--            <artifactId>quarkus-spring-web-api</artifactId>-->
+<!--        </dependency>-->
```
Comment (gsmet) [inline integration-tests/spring-scheduling/pom.xml:47]: This will need some cleanup.

**Review by gsmet — DISMISSED** [review]
Added some comments.

This will need a squash. Also, are the cron expression formats used by Spring and by our extension the same?

**Comment by aureamunoz** [comment]
It goes with https://github.com/quarkusio/quarkus-spring-api/pull/14

File: bom/runtime/pom.xml (hunk 1 above)
Comment (aureamunoz) [reply bom/runtime/pom.xml:741]: yes, ok. Thx!

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 4 above)
Comment (aureamunoz) [reply extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:59]: Of course not, I need to work on this more. It was just to put the basics in place and see if it works :-)

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 5 above)
Comment (aureamunoz) [reply extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:55]: I don't get what do you mean. Yes, this annotation is repeatable.

**Comment by aureamunoz** [comment]
> Added some comments.
> 
> This will need a squash. Also, are the cron expression formats used by Spring and by our extension the same?
Yes, a squash will be done. I have to check more in detail for cron expressions but from javadoc both seem the same.

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringScheduledAnnotatedMethodBuildItem.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import java.util.Map;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.MethodInfo;
+
+import io.quarkus.builder.item.SimpleBuildItem;
+
+/**
+ * Build Item recording all the methods that have been effectively annotated with
+ * {@code @Scheduled}
```
Comment (geoand) [inline extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringScheduledAnnotatedMethodBuildItem.java:12]:
```suggestion
 * {@code @org.springframework.scheduling.annotation.Scheduled}
```

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 5 above)
Comment (geoand) [reply extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:55]: @gsmet means that you will also have to handle `@Schedules`

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    @BuildStep
+    void registerBeans(SpringScheduledAnnotatedMethodBuildItem methodBuildItem,
+            BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
+        Map<MethodInfo, AnnotationInstance> methodToInstanceMap = methodBuildItem.getMethodToInstanceMap();
+        Set<String> beanClasses = new HashSet<>();
+        for (MethodInfo methodInfo : methodToInstanceMap.keySet()) {
+            beanClasses.add(methodInfo.declaringClass().asClass().name().toString());
+        }
+        additionalBeans.produce(AdditionalBeanBuildItem.builder().addBeanClasses(beanClasses).setUnremovable()
```
Comment (geoand) [inline extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:92]: This is a little weird TBH - Does regular Spring make a class a bean by simply annotating with `@Scheduled` or does it already have to be a bean (by having `@Component` for example)?

File: extensions/spring-scheduling/deployment/pom.xml (hunk 2 above)
Comment (aureamunoz) [reply extensions/spring-scheduling/deployment/pom.xml:32]: :+1:

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 8 above)
Comment (aureamunoz) [reply extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:92]: Ok. No, Spring does not make a class a bean by annotating with @Scheduled. It has to be already a bean in order to Spring can detect the method annotated with @Scheduled.

**Comment by aureamunoz** [comment]
Here there is another approach following what we were talking about. 
I search for Spring `@Scheduled` methods in beans and produce a `ScheduledBusinessMethodItem` with a regular Quarkus `@Scheduled` instance for each of them. `ScheduledBusinessMethodItem` is processed by the Quarkus scheduler extension.
Added also an integration test and squash the commits.
At the moment there is only the basic case of an annotation with a cron expression, I'm working in the other cases but wanted to have a review on this in order to be sure that you agree with it. I need of course add more tests.
WDYT @geoand ?

**Comment by geoand** [comment]
Yes, this make more sense. If the tests pass, then it's all good :)

**Comment by aureamunoz** [comment]
Hi,
I added other cases and unit tests and integration tests, I think it can be reviewed. I need to work on documentation but may be do it in a separate PR? WDYT @geoand @gsmet ?

File: bom/deployment/pom.xml [hunk 9]
```diff
@@ -748,6 +748,11 @@
                 <artifactId>quarkus-spring-cache-deployment</artifactId>
                 <version>${project.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-spring-scheduling-deployment</artifactId>
```
Comment (geoand) [inline bom/deployment/pom.xml:753]: This should probably be `scheduled` instead of `scheduling` because we use `quarkus-scheduled` in the existing extension.

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java [hunk 10]
```diff
@@ -106,6 +106,7 @@
     public static final String SPRING_BOOT_PROPERTIES = "spring-boot-properties";
     public static final String SPRING_CACHE = "spring-cache";
     public static final String SPRING_CLOUD_CONFIG_CLIENT = "spring-cloud-config-client";
+    public static final String SPRING_SCHEDULING = "spring-scheduling";
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java:109]: This should probably be `spring-scheduled`

File: extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.scheduler.deployment.ScheduledBusinessMethodItem;
+import io.quarkus.scheduler.runtime.SchedulerContext;
+
+/*
+ * A simple processor that maps annotations Spring Scheduled annotation to Quarkus annotation
+ * Arc's handling of annotation mapping (by creating an extra abstraction layer on top of the Jandex index)
+ * suits this sort of handling perfectly
```
Comment (geoand) [inline extensions/spring-scheduling/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:35]: The comment is kind of hard to read. Can you try and make it a little better?

File: extensions/spring-scheduling/deployment/src/test/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessorTest.java [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+import io.quarkus.arc.processor.BeanArchives;
+import io.quarkus.deployment.util.IoUtil;
+
+public class SpringSchedulingProcessorTest {
+
+    final SpringSchedulingProcessor springSchedulingProcessor = new SpringSchedulingProcessor();
+    final IndexView index = getIndex(SpringScheduledMethodsBean.class);
+
+    @Test
+    public void test_buildCronParam() {
```
Comment (geoand) [inline extensions/spring-scheduling/deployment/src/test/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessorTest.java:31]: We don't really use underscores, can you please stick to camel case?

**Comment by geoand** [comment]
Made a super quick first pass and it looks good, :+1: .

I still need to take a much closer look and I'll try to do that tomorrow, but realistically that will probably not happen, so please be patient for a few days :)

**Comment by aureamunoz** [comment]
Ok, thank you @geoand I will review your comments and work on documentation in the meantime

**Comment by geoand** [comment]
@aureamunoz, IIUC correctly from your Pull Request, Spring's ability to read the cron expression from `application.properties` using `${}` isn't supported, right? If not, is there a proper error message?

File: extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+    private long getLongValueFromParam(List<AnnotationValue> springAnnotationValues, String paramName) {
+        long paramValue = 0;
+        String paramValueString = "";
+        if (annotationsValuesContain(springAnnotationValues, paramName)) {
+            paramValue = getAnnotationValueByName(springAnnotationValues, paramName).get().asLong();
+        } else { //param value as String e.g. a placeholder ${value.from.conf}
+            paramValueString = getAnnotationValueByName(springAnnotationValues, paramName + "String")
+                    .get().asString();
+            paramValueString = paramValueString.replace("${", "{").trim();
```
Comment (geoand) [inline extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:160]: What does this do?

File: extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+                                || annotationsValuesContain(springAnnotationValues, "initialDelayString")) {
+                            confValues.addAll(buildDelayParams(springAnnotationValues));
+                        }
+
+                    } else if (annotationsValuesContain(springAnnotationValues, "fixedDelay")) {
+                        throw new IllegalStateException(
+                                "Unsupported @Scheduled param fixedDelay");
+                    } else if (annotationsValuesContain(springAnnotationValues, "cron")) {
+                        if (annotationsValuesContain(springAnnotationValues, "initialDelay")) {
+                            throw new IllegalStateException(
```
Comment (geoand) [inline extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:109]: We should have a consistent format for the error messages.  Also, we should probable use `IllegalArgumentException` instead of `IllegalStateException`

File: extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+            if (schedules != null) {
+                scheduledBusinessMethods.produce(new ScheduledBusinessMethodItem(bean, method, schedules));
+                LOGGER.debugf("Found scheduled business method %s declared on %s", method, bean);
+            }
+        }
+    }
+
+    AnnotationValue buildCronParam(List<AnnotationValue> springAnnotationValues) {
+        String cronExpression = getAnnotationValueByName(springAnnotationValues, "cron").get().asString();
+        cronExpression = cronExpression.replace("${", "{");
```
Comment (aureamunoz) [inline extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:130]: > @aureamunoz, IIUC correctly from your Pull Request, Spring's ability to read the cron expression from `application.properties` using `${}` isn't supported, right? If not, is there a proper error message?
Yes @geoand , it is supported.

Comment (geoand) [reply extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:130]: Is there a test that shows this functionality? I mean adding a cron expression in `application.properties` and making sure that it works.

File: extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 13 above)
Comment (aureamunoz) [reply extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:160]: It reads the value from `fixedRateString` and `initialDelayString` params in Spring `@Scheduled`. This params are used to read values from `application.properties`

Comment (geoand) [reply extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:160]: I see, thanks! Do we have tests to verify both kinds of behavior?

Also, we need to note that in Quarkus this support is for build-time properties, not runtime properties

File: extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 15 above)
Comment (aureamunoz) [reply extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:130]: Yes, in `SpringScheduledMethodTest.java` I add the cron expression in `application.properties` and have a test. And `SpringScheduledMethodsBean#checkEverySecondCronConfig` is the method.

File: extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 13 above)
Comment (aureamunoz) [reply extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java:160]: Yes, we have tests for both of them.
Ok, I note it for the documentation.

File: extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java (hunk 14 above)
Comment (aureamunoz) [reply extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java]: I changed the exception type and homogenized the message, wdyt @geoand

Comment (geoand) [reply extensions/spring-scheduled/deployment/src/main/java/io/quarkus/spring/scheduling/deployment/SpringSchedulingProcessor.java]: :+1:

**Comment by geoand** [comment]
I think this is good, but I'll need to take a closer look towards the end of the week,

File: extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/SimpleJobs.java [hunk 16]
```diff
@@ -4,8 +4,11 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 
+import javax.enterprise.context.ApplicationScoped;
+
 import io.quarkus.scheduler.Scheduled;
 
+@ApplicationScoped
```
Comment (geoand) [inline extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/SimpleJobs.java:11]: Hm, why did you need to add this?

**Comment by geoand** [comment]
I think the integration test should be moved to `spring-web`, instead of introducing another Maven module. That way you would get native image testing for free (assuming you add the IT test like we have for others).
Also you need a file like this: https://github.com/quarkusio/quarkus/blob/778ff5196edbfff83faf6eaf926e05c2f0030d6d/extensions/spring-cache/runtime/src/main/resources/META-INF/quarkus-extension.yaml

**Comment by aureamunoz** [comment]
> I think the integration test should be moved to `spring-web`, instead of introducing another Maven module. That way you would get native image testing for free (assuming you add the IT test like we have for others).
> Also you need a file like this: https://github.com/quarkusio/quarkus/blob/778ff5196edbfff83faf6eaf926e05c2f0030d6d/extensions/spring-cache/runtime/src/main/resources/META-INF/quarkus-extension.yaml

     

The quarkus-extension.yaml file is already present. It just needs a refactor after the renaming to spring-scheduled, I will do it.
OK, I will take a look for the IT.

File: extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/SimpleJobs.java (hunk 16 above)
Comment (aureamunoz) [reply extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/SimpleJobs.java:11]: It has to be a bean in order to detect the method annotated with @scheduled

**Comment by geoand** [comment]
Ah yeah, I missed the YAML file. But the integration test should be taken care of

File: extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/SimpleJobs.java (hunk 16 above)
Comment (geoand) [reply extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/SimpleJobs.java:11]: But why did you have to add it as part of the PR?

How did it work before?

Comment (aureamunoz) [reply extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/SimpleJobs.java:11]: Ah, sorry, my bad. I thought I was watching at SimpleJobs.java in spring-scheduled extension. I should have added it to test something and it slipped through the cracks :-S Let me remove it

**Review by geoand — APPROVED** [review]
LGTM, thanks a lot!

File: extensions/spring-scheduled/deployment/pom.xml [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    <name>Quarkus - Spring - Scheduling - Deployment</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-spring-scheduled</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-core-deployment</artifactId>
```
Comment (geoand) [inline extensions/spring-scheduled/deployment/pom.xml:24]: You definitely don't need this dependency since it's brought in transitively. But this can be fixed in a follow up.

File: extensions/spring-scheduled/deployment/pom.xml [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc-deployment</artifactId>
+        </dependency>
```
Comment (geoand) [inline extensions/spring-scheduled/deployment/pom.xml:29]: I am pretty sure that this is also brought in by `quarkus-scheduler-deployment`. Please check and fix in a follow up.

**Comment by aureamunoz** [comment]
Thank you @geoand !

**Comment by geoand** [comment]
Thank you for your work!
